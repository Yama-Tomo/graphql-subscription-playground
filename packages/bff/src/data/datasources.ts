import { Context } from '@/context';

type Options = { cache?: boolean };
type DataSourceContext = Omit<Context, 'dataSources'>;

export type { Options, DataSourceContext };
