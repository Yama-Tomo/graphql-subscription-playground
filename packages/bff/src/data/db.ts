import * as fs from 'fs';
import * as path from 'path';

type Message = {
  channelId: string;
  id: string;
  text: string;
  userId: string;
  date: Date;
  readUserIds: string[];
};

type Channel = {
  description?: string | null;
  id: string;
  isDM: boolean;
  joinUserIds: string[];
  name: string;
  ownerId: string;
};

type ChannelWithPersonalizedData = Channel;

type User = {
  id: string;
  name: string;
};

type ReadMessageUsers = {
  id: string;
  readUserIds: string[];
};

type DB = { messages: Message[]; channels: Channel[]; users: User[] };

const db: DB = (() => {
  if (process.env.NODE_ENV === 'test') {
    throw new Error('DB values must be mocked');
  }

  const dataFile = path.resolve(__dirname, 'db.json');
  const flushToDisk = () => fs.writeFileSync(dataFile, JSON.stringify(db));

  process.once('SIGUSR2', function () {
    flushToDisk();
    process.kill(process.pid, 'SIGUSR2');
  });

  process.on('SIGINT', function () {
    flushToDisk();
  });

  if (fs.existsSync(dataFile)) {
    try {
      return JSON.parse(fs.readFileSync(dataFile, 'utf8'));
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }

  const messages: Message[] = [];
  const channels: Channel[] = [];
  const users: User[] = [];
  return { messages, channels, users };
})();

export { db };
export type { DB, Channel, Message, User, ChannelWithPersonalizedData, ReadMessageUsers };
