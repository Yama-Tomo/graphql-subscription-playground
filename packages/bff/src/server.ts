import { ApolloServer } from 'apollo-server-fastify';
import { ApolloServerPluginDrainHttpServer } from 'apollo-server-core';
import { fastify } from 'fastify';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { execute, subscribe } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';
import { typeDefs } from '@/schema';
import { resolvers } from '@/resolvers';
import { UnAuthorizedContext, createAllRequestSharedContext } from '@/context';

const startServer = async (opts?: Partial<{ port: number; path: string }>) => {
  const path = opts?.path || '/graphql';
  const fastifyApp = fastify();
  const sharedContext = createAllRequestSharedContext();

  const schema = makeExecutableSchema({ typeDefs, resolvers });

  const subscriptionServer = SubscriptionServer.create(
    {
      schema,
      execute,
      subscribe,
      onConnect(params: Record<string, unknown>): UnAuthorizedContext {
        const userId = params['user_id'] ? String(params['user_id']) : undefined;
        if (!userId) {
          throw new Error('unauthorized');
        }

        return { ...sharedContext, user: { id: userId } };
      },
    },
    { server: fastifyApp.server, path: path }
  );

  const server = new ApolloServer({
    schema,
    context(param): UnAuthorizedContext {
      const userId = param.request.headers['user_id']
        ? String(param.request.headers['user_id'])
        : undefined;
      return { ...sharedContext, ...(userId ? { user: { id: userId } } : {}) };
    },
    plugins: [
      ApolloServerPluginDrainHttpServer({ httpServer: fastifyApp.server }),
      {
        async serverWillStart() {
          return {
            async drainServer() {
              subscriptionServer.close();
            },
          };
        },
      },
    ],
  });

  await server.start();
  fastifyApp.register(server.createHandler({ path }));
  const listenHost = await fastifyApp.listen(opts?.port || 4000);

  return { server, fastifyApp, publicUrl: listenHost + path };
};

export { startServer };
