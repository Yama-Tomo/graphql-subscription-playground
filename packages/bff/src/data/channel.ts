import { DataSource, DataSourceConfig } from 'apollo-datasource';
import DataLoader from 'dataloader';
import { v4 } from 'uuid';

import { DataSourceContext, Options } from '@/data/datasources';
import { Channel, db } from '@/data/db';

class ChannelDataSource extends DataSource {
  context!: DataSourceContext;
  private byIdLoader: DataLoader<string, Channel>;

  constructor(options?: Options) {
    super();

    this.byIdLoader = new DataLoader(
      async (ids) => {
        return ids.map(
          (id) => db.channels.find((v) => v.id === id) || new Error(`channel not found: [${id}]`)
        );
      },
      { cache: options?.cache ?? true }
    );
  }

  initialize(config: DataSourceConfig<DataSourceContext>) {
    this.context = config.context;
  }

  async getById(id: string) {
    return this.byIdLoader.load(id);
  }

  async getByIds(ids: string[]) {
    const items = await this.byIdLoader.loadMany(ids);

    const error = items.find((item): item is Error => item instanceof Error);
    if (error) {
      throw error;
    }

    return items as Channel[];
  }

  getJoinedChannels(userId: string) {
    return db.channels.filter((channel) => channel.joinUserIds.includes(userId));
  }

  create(data: Omit<Channel, 'id'>) {
    const channel = { ...data, id: v4() };
    db.channels.push(channel);

    return channel;
  }

  async update(data: Pick<Channel, 'id'> & Partial<Omit<Channel, 'id'>>) {
    const channel = await this.getById(data.id);
    const { joinUserIds, ...rest } = data;

    (Object.keys(rest) as (keyof typeof rest)[]).forEach((key) => {
      if (channel[key] !== data[key]) {
        (channel[key] as unknown) = data[key];
      }
    });

    if (JSON.stringify(joinUserIds) !== JSON.stringify(data.joinUserIds)) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      channel.joinUserIds = data.joinUserIds!;
    }

    return channel;
  }

  delete(id: Channel['id']) {
    const dataIdx = db.channels.findIndex((channel) => channel.id === id);
    const channel = db.channels[dataIdx];
    if (!channel) {
      throw new Error('channel not found');
    }

    db.channels.splice(dataIdx, 1);

    return channel;
  }

  async addUser(id: Channel['id'], userId: Channel['joinUserIds'][number]) {
    const channel = await this.getById(id);

    channel.joinUserIds.push(userId);

    return channel;
  }
}

export { ChannelDataSource };
