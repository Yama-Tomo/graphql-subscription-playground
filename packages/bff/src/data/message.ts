import { DataSource, DataSourceConfig } from 'apollo-datasource';
import DataLoader from 'dataloader';
import { v4 } from 'uuid';

import { DataSourceContext, Options } from '@/data/datasources';
import { db, Message } from '@/data/db';

class MessageDataSource extends DataSource {
  context!: DataSourceContext;
  private byIdLoader: DataLoader<string, Message>;
  private byChannelIdLoader: DataLoader<string, Message[]>;

  constructor(options?: Options) {
    super();

    const cache = options?.cache ?? true;
    this.byIdLoader = new DataLoader(
      async (ids) => {
        return ids.map(
          (id) => db.messages.find((v) => v.id === id) || new Error(`message not found: [${id}]`)
        );
      },
      { cache }
    );

    this.byChannelIdLoader = new DataLoader<string, Message[]>(
      async (ids) => {
        return ids.map((id) => db.messages.filter((v) => v.channelId === id));
      },
      { cache }
    );
  }

  initialize(config: DataSourceConfig<DataSourceContext>) {
    this.context = config.context;
  }

  async getById(id: string) {
    return this.byIdLoader.load(id);
  }

  async getByIds(ids: string[]) {
    const items = await this.byIdLoader.loadMany(ids);

    const error = items.find((item): item is Error => item instanceof Error);
    if (error) {
      throw error;
    }

    return items as Message[];
  }

  getByChannelId(id: string) {
    return this.byChannelIdLoader.load(id);
  }

  create(data: Omit<Message, 'id'>) {
    const message = { ...data, id: v4() };
    db.messages.push(message);

    return message;
  }

  async update(data: Pick<Message, 'id'> & Partial<Omit<Message, 'id'>>) {
    const message = await this.getById(data.id);
    const { readUserIds, ...rest } = data;

    console.log(data);
    (Object.keys(rest) as (keyof typeof rest)[]).forEach((key) => {
      if (message[key] !== data[key]) {
        (message[key] as unknown) = data[key];
      }
    });

    if (data.readUserIds && JSON.stringify(readUserIds) !== JSON.stringify(data.readUserIds)) {
      message.readUserIds = data.readUserIds;
    }

    return message;
  }

  delete(id: Message['id']) {
    const dataIdx = db.messages.findIndex((channel) => channel.id === id);
    const message = db.messages[dataIdx];
    if (!message) {
      throw new Error('message not found');
    }

    db.messages.splice(dataIdx, 1);

    return message;
  }
}

export { MessageDataSource };
