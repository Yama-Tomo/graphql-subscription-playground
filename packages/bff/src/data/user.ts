import { DataSource, DataSourceConfig } from 'apollo-datasource';
import DataLoader from 'dataloader';
import { db, User } from '@/data/db';
import { DataSourceContext, Options } from '@/data/datasources';
import { v4 } from 'uuid';

class UserDataSource extends DataSource {
  context!: DataSourceContext;
  private byIdLoader: DataLoader<string, User>;

  constructor(options?: Options) {
    super();

    this.byIdLoader = new DataLoader(
      async (ids) => {
        return ids.map(
          (id) => db.users.find((v) => v.id === id) || new Error(`user not found: [${id}]`)
        );
      },
      { cache: options?.cache ?? true }
    );
  }

  initialize(config: DataSourceConfig<DataSourceContext>) {
    this.context = config.context;
  }

  async getById(id: string) {
    return this.byIdLoader.load(id);
  }

  async getByIds(ids: string[]) {
    const items = await this.byIdLoader.loadMany(ids);

    const error = items.find((item): item is Error => item instanceof Error);
    if (error) {
      throw error;
    }

    return items as User[];
  }

  getByName(name: string) {
    // キーワード検索のように検索条件が幅広い場合はキャッシュせずにクライアント側のキャッシュ戦略に任せたほうが良い気がする
    return db.users.filter((usr) => usr.name.includes(name));
  }

  create(data: Omit<User, 'id'>) {
    const user = { ...data, id: v4() };
    db.users.push(user);

    return user;
  }
}

export { UserDataSource };
