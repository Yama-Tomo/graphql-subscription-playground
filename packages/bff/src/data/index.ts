import { Config } from 'apollo-server-core';
import { MessageDataSource } from '@/data/message';
import { UserDataSource } from '@/data/user';
import { DataSourceContext, Options } from '@/data/datasources';
import { ChannelDataSource } from '@/data/channel';

type Cache = NonNullable<Config['cache']>;

const dataSources = (options?: Options) => ({
  message: new MessageDataSource(options),
  channel: new ChannelDataSource(options),
  user: new UserDataSource(options),
});

const initDataSources = async (context: DataSourceContext, cache: Cache, options?: Options) => {
  const ds = dataSources(options);
  const initializers: Array<void | Promise<void>> = Object.values(ds).map((dsInstance) => {
    return dsInstance.initialize({ context, cache });
  });

  await Promise.all(initializers);
  return ds;
};

type DataSources = ReturnType<typeof dataSources>;

export { dataSources, initDataSources };
export * from './db';
export type { DataSources };
