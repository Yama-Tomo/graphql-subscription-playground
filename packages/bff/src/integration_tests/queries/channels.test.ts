import { FastifyReply, FastifyRequest } from 'fastify';

import { DB } from '@/data';
import { startServer } from '@/server';
import { mockDBData } from '@/test_utils/mocks';

test('メッセージ一覧', async () => {
  const user1 = { id: 'u1', name: 'user1' };
  const user2 = { id: 'u2', name: 'user2' };
  const dbData: DB = {
    // prettier-ignore
    messages: [
      { id: 'mes1', readUserIds: [user1.id], userId: user1.id, channelId: 'ch1', text: 'mes1', date: new Date('2021-01-01') },
      { id: 'mes2', readUserIds: [], userId: user1.id, channelId: 'ch1', text: 'mes2', date: new Date('2021-01-01') },
      { id: 'mes3', readUserIds: [], userId: user2.id, channelId: 'ch1', text: 'mes3', date: new Date('2021-01-01') },
    ],
    users: [user1, user2],
    // prettier-ignore
    channels: [
      { id: 'ch1', name: 'ch1', description: '', joinUserIds: [user1.id, user2.id], isDM: false, ownerId: user1.id },
    ],
  };

  mockDBData(dbData);

  const { server } = await startServer();
  const query = `query ($channelId: ID!, $last: Int!) {
    messages(channelId: $channelId, last: $last) {
      pageInfo {
        endCursor
        startCursor
        hasPreviousPage
        hasNextPage
      }
      edges {
        cursor
        node {
          id
          user {
            id
            name
          }
          readUsers {
            id
            name
          }
          date
          text
          channelId
        }
      }
    }
  }`;
  const res = await server.executeOperation(
    { query, variables: { channelId: 'ch1', last: 10 } },
    {
      request: { headers: { user_id: user1.id } } as unknown as FastifyRequest,
      reply: {} as FastifyReply,
    }
  );

  expect(res).toMatchSnapshot();
});
