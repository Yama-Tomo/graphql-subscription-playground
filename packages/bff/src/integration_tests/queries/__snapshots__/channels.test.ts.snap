// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`メッセージ一覧 1`] = `
Object {
  "data": Object {
    "messages": Object {
      "edges": Array [
        Object {
          "cursor": "mes1",
          "node": Object {
            "channelId": "ch1",
            "date": 2021-01-01T00:00:00.000Z,
            "id": "mes1",
            "readUsers": Array [
              Object {
                "id": "u1",
                "name": "user1",
              },
            ],
            "text": "mes1",
            "user": Object {
              "id": "u1",
              "name": "user1",
            },
          },
        },
        Object {
          "cursor": "mes2",
          "node": Object {
            "channelId": "ch1",
            "date": 2021-01-01T00:00:00.000Z,
            "id": "mes2",
            "readUsers": Array [],
            "text": "mes2",
            "user": Object {
              "id": "u1",
              "name": "user1",
            },
          },
        },
        Object {
          "cursor": "mes3",
          "node": Object {
            "channelId": "ch1",
            "date": 2021-01-01T00:00:00.000Z,
            "id": "mes3",
            "readUsers": Array [],
            "text": "mes3",
            "user": Object {
              "id": "u2",
              "name": "user2",
            },
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "mes3",
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": "mes1",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
